packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -61,40 +62,116 @@ export interface IBlobManagerLoadInfo {
+    OfflinePendingUpload,
+    OfflinePendingOp,
+}
+
+interface PendingBlob {
+    blob: ArrayBufferLike;
+    status: PendingBlobStatus;
+    storageId?: string;
+    // resolved when the online flow is complete (i.e. BlobAttach op round-tripped) or when we
+    // disonnect and transition everything to offline flow
+    deferred: Deferred<string>;
+}
+
 export class BlobManager {
     public static readonly basePath = "_blobs";
     private static readonly redirectTableBlobName = ".redirectTable";
-    // uploaded blob IDs
-    private readonly blobIds: Set<string> = new Set();
-    // blobs for which upload is pending. maps to a promise that will resolve once the blob has been uploaded and a
-    // BlobAttach op has round-tripped.
-    private readonly pendingBlobIds: Map<string, Deferred<void>> = new Map();
-    // blobs uploaded while detached; cleared upon attach
-    private readonly detachedBlobIds: Set<string> = new Set();
-    // map of detached blob IDs to IDs used by storage. used to support blob handles given out while detached
-    private redirectTable: Map<string, string> | undefined;
+    private readonly logger: ITelemetryLogger;
+
+    // Map of local (offline/detached) IDs to storage IDs. Contains identity entries
+    // (id -> id) for storage IDs, so all requested IDs should be a key in this map.
+    private readonly redirectTable: Map<string, string | undefined>;
+
+    // Blobs which have not been uploaded or for which we have not yet seen a BlobAttach op round-trip
+    private readonly pendingBlobs: Map<string, PendingBlob> = new Map();
+
+    // Map of storage IDs to list of local IDs waiting for the BlobAttach op with that storage ID.
+    // Note this doesn't include "offline" BlobAttach ops since they will be 1:1 with "offline" blobs.
+    private readonly opsInFlight: Map<string, string[]> = new Map();
 
     constructor(
         private readonly routeContext: IFluidHandleContext,
         snapshot: IBlobManagerLoadInfo,
         private readonly getStorage: () => IDocumentStorageService,
-        private readonly attachBlobCallback: (blobId: string) => void,
+        private readonly sendBlobAttachOp: (blobId?: string, localId?: string) => void,
         // To be called when a blob node is requested. blobPath is the path of the blob's node in GC's graph. It's
         // of the format `/<BlobManager.basePath>/<blobId>`.
         private readonly gcNodeUpdated: (blobPath: string) => void,
-        private readonly runtime: IContainerRuntime,
-        private readonly logger: ITelemetryLogger,
+        private readonly runtime: IBlobManagerRuntime,
     ) {
-        this.runtime.once("dispose", () => {
-            for (const promise of this.pendingBlobIds.values()) {
-                promise.reject(new Error("runtime disposed while blobAttach op in flight"));
+        this.logger = ChildLogger.create(this.runtime.logger, "BlobManager");
+        this.runtime.on("disconnected", () => this.onDisconnected());
```

Comment:
It's a bit hard to provide feedback without a design doc, but I wonder if blob manager should not know anything about connected and disconnected events at all. I.e. what if the flow is the following:
- Blob manager always returns a handle synchronously on blob creation.
- It attempts to upload blob, infinitely retrying on any retryable error (which includes offline mode).
- Once upload completed, it submits an op
- If op ack times out, it starts over.

This feels like make sequence really simple / straightforward, make almost no distinction between offline and online, which ensures it's easier to test / we have less bugs.
We could add capability to throttle these calls not to burn CPU while we are offline by providing a layer that pauses this network calls (does not let them through) while client is disconnected. But I'd do it as an optional layer, not baked into core of batch manager.

Reply:
The options we've considered are:

a)  Return IFluidHandles asynchronously, permitting large blobs to upload concurrently with collaboration.
b)  Return IFluidHandles synchronously while:
    1.  blocking outgoing ops as large blobs upload.
    2.  allowing outgoing ops, which allows peers to obtain references to handles that will never resolve.

Our current strategy is to use 'A' when online, which allows applications to upload large blobs "in the background" while continuing to collaborate.

We can't use 'A' when working offline, so we've chosen to accept B1, as it seems less hazardous for the end user and blocking collaboration while uploading the user's offline work also feels "more okay" than blocking collaboration after connecting.

I don't love the asymmetry, but the choice seems rationale.